src/index.ts

```diff
@@ -0,0 +1,312 @@
+    const first3CharactersChanged = !(
+      text.substr(0, shortenedPrevQ.length) === shortenedPrevQ
+    );
+    const queryCleared = this.lastQuery.length > 0 && text.length === 0;
+    if (first3CharactersChanged || queryCleared) {
+      this.reset();
+    }
+    this.lastQuery = text;
+
+    return this.session.next(values);
+  }
+
+  public reset(): void {
+    this.session.reset();
+  }
+}
+
+export class Session implements ISession {
+  private queryID: string = "";
+  private sequence: number = 0;
+
+  private trackingType: TrackingType;
+  private field: string;
+  private sessionData: Values;
+
+  public constructor(trackingType: TrackingType, field: string, data: Values) {
+    this.trackingType = trackingType;
+    this.field = field;
+    this.sessionData = data;
+  }
+
+  public next(values: Values): [Tracking | undefined, Error | undefined] {
+    if (this.queryID === "") {
+      this.queryID = randString();
+      this.sequence = 0;
+    } else {
+      this.sequence++;
+    }
+
+    return [
+      {
+        type: this.trackingType,
+        query_id: this.queryID,
+        sequence: this.sequence,
+        field: this.field,
+        data: this.sessionData
+      },
+      undefined
+    ];
+  }
+
+  public reset(): void {
+    this.queryID = "";
+    this.sequence = 0;
+  }
+}
+
+export type Opt = (client: Client) => void;
+
+export const withEndpoint = (endpoint: string) => (client: Client) => {
+  client.endpoint = endpoint;
+};
+
+/**
+ * Client takes configuration and constructs pipelines clients
+ */
+export class Client {
+  public project: string;
+  public collection: string;
+  public endpoint: string;
+
+  public constructor(project: string, collection: string, opts: Opt[] = []) {
+    this.project = project;
+    this.collection = collection;
+    this.endpoint = "https://jsonapi.sajari.net/";
+    opts.forEach(opt => {
+      opt(this);
+    });
+  }
+
+  /**
+   * pipeline returns a new pipeline client
+   */
+  public pipeline(name: string): Pipeline {
+    return new Pipeline(this, name);
+  }
+}
+
+export interface SJError {
+  message: string;
+  code?: number;
+}
+
+const makeError = (message: string, code?: number): SJError => ({
+  message,
+  code
+});
+
+export interface Key {
+  field: string;
+  value: any;
+}
+
+export type SearchCallback = (
+  error?: SJError,
+  results?: Results,
+  values?: Values
+) => void;
+
+export type AddCallback = (key: Key, error: SJError) => void;
+
+export interface SJRecord {
+  [id: string]: any;
+}
+
+/**
+ * makeRequest makes a XMLHttpRequest and handles network and parsing errors.
+ */
+const makeRequest = (
+  address: string,
+  body: any,
+  callback: (error?: SJError, response?: any) => void
+): void => {
+  const request = new XMLHttpRequest();
+  request.open("POST", address, true);
+  request.setRequestHeader("Accept", "application/json");
+  request.onreadystatechange = () => {
+    if (request.readyState !== 4) return;
```

Comment:
I know I added one of these magic numbers below, but it might be worth creating a const for each of them.